File: src/main/java/ukalus/level/maze/kruskal/KruskalMazeGenerator.java
package ukalus.level.maze.kruskal;

import org.apache.commons.math3.random.MersenneTwister;
import org.apache.commons.math3.random.RandomAdaptor;
import ukalus.level.Level;
import ukalus.level.Region;
import ukalus.level.RegionFactory;
import ukalus.level.maze.MazeRegion;
import ukalus.math.Vector2D;
import ukalus.util.DisjointSet;
import ukalus.util.Loop;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class KruskalMazeGenerator implements RegionFactory {

  private Random random;
  private int height;
  private int width;

  public KruskalMazeGenerator(Random random, int height, int width) {
    this.random = random;
    this.height = ((height - 1) & (Integer.MAX_VALUE - 1)) + 1;
    this.width = ((width - 1) & (Integer.MAX_VALUE - 1)) + 1;
  }

  public Region create() {
    int[][] nodes = new int[height][width];
    List<EdgeCell> edges = new ArrayList<EdgeCell>(height * width);

    // add walls / open cells
    for (int x = 1; x < nodes.length - 1; x++) {
      for (int y = 1; y < nodes[x].length - 1; y++) {
        if (x % 2 == 0 && y % 2 == 1) {
          // horizontal wall
          edges.add(new EdgeCell(x, y, (x - 1) * width + y, (x + 1) * width + y));
        } else if (x % 2 == 1 && y % 2 == 0) {
          // vertical wall
          edges.add(new EdgeCell(x, y, x * width + y - 1, x * width + y + 1));
        } else if (x % 2 == 1 && y % 2 == 1) {
          // open cell
          nodes[x][y] = 1;
        }
      }
    }

    Collections.shuffle(edges, random);
    new Loop<EdgeCell>(edges).forEach(new TraverseEdge(new DisjointSet(height
        * width), nodes));

    return new MazeRegion(nodes);
  }

  public static void main(String[] args) {
    RegionFactory generator = new KruskalMazeGenerator(new RandomAdaptor(new MersenneTwister()),
      20, 80);

    Level level = new Level(new Object[20][80]);
    Region region = generator.create();
    region.place(Vector2D.get(0, 0), level);

    System.out.println(level);
  }

}

File: src/main/java/ukalus/level/maze/kruskal/KruskalMazeGenerator.kt
package ukalus.level.maze.kruskal

import org.apache.commons.math3.random.MersenneTwister
import org.apache.commons.math3.random.RandomAdaptor
import ukalus.level.Level
import ukalus.level.Region
import ukalus.level.RegionFactory
import ukalus.level.maze.MazeRegion
import ukalus.util.shuffle
import ukalus.math.Vector2D
import ukalus.util.DisjointSet
import java.util.*

class KruskalMazeGenerator(private val random: Random, height: Int, width: Int) : RegionFactory<Int> {
    private val height: Int = (height - 1 and Integer.MAX_VALUE - 1) + 1
    private val width: Int = (width - 1 and Integer.MAX_VALUE - 1) + 1

    override fun create(): Region<Int> {
        val nodes = Array(height) { IntArray(width) }
        val edges = ArrayList<EdgeCell>(height * width)

        // add walls / open cells
        for (x in 1..nodes.size - 1 - 1) {
            for (y in 1..nodes[x].size - 1 - 1) {
                if (x % 2 == 0 && y % 2 == 1) {
                    // horizontal wall
                    edges.add(EdgeCell(x, y, (x - 1) * width + y, (x + 1) * width + y))
                } else if (x % 2 == 1 && y % 2 == 0) {
                    // vertical wall
                    edges.add(EdgeCell(x, y, x * width + y - 1, x * width + y + 1))
                } else if (x % 2 == 1 && y % 2 == 1) {
                    // open cell
                    nodes[x][y] = 1
                }
            }
        }

        val sets = DisjointSet(height * width)
        edges
                .shuffle(random)
                .forEach { wall ->
                    if (sets.find(wall.left) != sets.find(wall.right)) {
                        sets.union(wall.left, wall.right)
                        nodes[wall.x][wall.y] = 1
                    }
                }

        return MazeRegion(nodes)
    }

    companion object {

        @JvmStatic fun main(args: Array<String>) {
            val generator = KruskalMazeGenerator(RandomAdaptor(MersenneTwister()), 19, 79)

            val level = Level(Array(19) { Array(79, { 0 }) })
            val region = generator.create()
            region.place(Vector2D(0, 0), level)

            println(level)
        }
    }

}
